File: Templates/Templates/ros2_templates/01_Topic_Templates/README_Node.cpp
//  This document describes syntax and template for a ros2 node
// Created within the src folder
// Remember ros2 cpp -> build, soruce, run

// See this: https://docs.ros.org/en/jazzy/p/rclcpp/
// See This: https://github.com/ros2/rclcpp

// Topics use .msg


// ----------- Code starts here -----------
#include "rclcpp/rclcpp.hpp"         // Including the Ros Clinet Library API for ROS2

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);        // Initialises the ROS2 communication
    
    // ROS2 uses pointers and smart pointers for everything
    auto node = std::make_shared<rclcpp::Node>("node_name");     // Create a share pointer to a node object

    RCLCPP_INFO(node->get_logger(), "Hello World"); 

    rclcpp::spin(node);              // Adds the spin function
    rclcpp::shutdown();              // Shutdown the ros2 node
    return 0;
}

// How to build the code
// Go to CMakeLists.txt -> 

// cmake_minimum_required(VERSION 3.8)
// project(my_cpp_package)

// if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
//   add_compile_options(-Wall -Wextra -Wpedantic)
// endif()

// # find dependencies
// find_package(ament_cmake REQUIRED)
// find_package(rclcpp REQUIRED)

// ---------- Adding executables ----------

// add_executable(<exe name>, src/<node file.cpp>)
// ament_target_dependencies(<exe name> rclcpp <other dependencies ...>)                   // Adding the dependencies

// add_executable(<other name>, src/<node file.cpp>)
// ament_target_dependencies(<other name> rclcpp <other dependencies ...>)                   // Adding the dependencies

// install(TARGES
//  <exe name>
//  <other exe>
//  DESTINATION lib/${PROJECT_NAME}                                                        // Installing exe in lib folder
// )


// ---------- This may be removed for cleaner build as long as it is in package.xml ----------
// if(BUILD_TESTING)
//  find_package(ament_lint_auto REQUIRED)
//  # the following line skips the linter which checks for copyrights
//  # comment the line when a copyright and license is added to all source files
//  set(ament_cmake_copyright_FOUND TRUE)
//  # the following line skips cpplint (only works in a git repo)
//  # comment the line when this package is in a git repo and when
//  # a copyright and license is added to all source files
//  set(ament_cmake_cpplint_FOUND TRUE)
//  ament_lint_auto_find_test_dependencies()
// endif()
// -------------------------------------------------------------------------------------------

// ament_package()

// After this colcon build can be run

// ---------- Template Cpp ROS2 with classes ----------
#include "rclcpp/rclcpp.hpp"                                                 // Including the Ros Clinet Library API for ROS2

class MyNode : public rclcpp::Node
{
public:
    MyNode() : Node("cpp_test"), counter_(0)
    {
        RCLCPP_INFO(this->get_logger(), "Node Initialized");                      // "this" is a pointer to the current object instance of MyNode
        
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&MyNode::timerCallback, this)
        );
    }

private:
    void timerCallback()
    {
        RCLCPP_INFO(this->get_logger(), "Hello  %d", counter_);
        counter_++;
    }
    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);        // Initialises the ROS2 communication
    
    // ROS2 uses pointers and smart pointers for everything
    auto node = std::make_shared<MyNode>();                           // Create a share pointer to a node object

    rclcpp::spin(node);                                                      // Adds the spin function
    rclcpp::shutdown();                                                      // Shutdown the ros2 node
    return 0;
}


// Only prints text to its own terminal window. It does not broadcast any data across the ROS2 network. Other nodes cannot see or hear it.
// Therefore this next example includes the core ros2 functionality of publish/subscribe

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp" // Required for std_msgs::msg::String

class MyNode : public rclcpp::Node
{
public:
    MyNode() : Node("cpp_test"), counter_(0)
    {
        RCLCPP_INFO(this->get_logger(), "Hello ROS2");

        // ---------- Publisher ----------
        publisher_ = this->create_publisher<std_msgs::msg::String>("example_topic", 10);

        // ---------- Subscriber ----------
        subscription_ = this->create_subscription<std_msgs::msg::String>(
            "example_topic",
            10,
            std::bind(&MyNode::listenerCallback, this, std::placeholders::_1)
        );

        // ---------- Timer ----------
        timer_ = this->create_wall_timer(
            std::chrono::seconds(1),
            std::bind(&MyNode::timerCallback, this)
        );
    }

private:
    void timerCallback()
    {
        auto msg = std_msgs::msg::String();
        msg.data = "Hello ROS2 Network: " + std::to_string(counter_);
        
        publisher_->publish(msg);

        RCLCPP_INFO(this->get_logger(), "Hello");
        counter_++;
    }

    void listenerCallback(const std_msgs::msg::String::SharedPtr msg)
    {
        RCLCPP_INFO(this->get_logger(), "Subscribed Listener Heard: '%s'", msg->data.c_str());
    }

    // Member variables
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
    int counter_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    
    auto node = std::make_shared<MyNode>();

    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

// using namsepace is a good tool fo rccp saving space

File: Templates/Templates/ros2_templates/01_Topic_Templates/README_Node.py
# This document describes syntax and template for a ros2 node
# Remember python code is made in the same folder as __init__.py file
# chmod +x makes it an executable
# 

# Topics use .msg

# See this: https://docs.ros.org/en/jazzy/p/rclpy/
# See this: https://github.com/ros2/rclpy

# ----------- Code starts here -----------

#!/usr/bin/env python3 <- this is a shebang line, it tells the system that this file should be run with python3 interpreter
# remember shebang in top of document always
import rclpy 
from rclpy.node import Node               # only one import pr. line is allowed

def main(args=None):
    rclpy.init(args=args)                 # Initialize the ROS2 communication
    node = Node("my_test")                # See that the node is here, not the python file itself
    node.get_logger().info("Hello ROS2")  # Log a message to the console
    rclpy.spin(node)                       # Keeps the node alive until Ctrl+C
    rclpy.shutdown()                      # Shutdown the ROS2 communication

if __name__ == "__main__":                # Checks if the script is being executed directly from the terminal rather than being imported as a module by another script.
    main()                                # Executes the main function when the script is run directly.

# In the terminal after running it you would get a: [info][timestamp][node_name] corresponding to line 13 and 14
# Important after making a node add it to the "Entry points" in the setup.py
# Remember that colcon build will place the executables in install folder
# Easy to get confused because the node is defined on line 13, but it does not need to have the same name as the package, and also in setup.py the node_name can be called something else. Good place to troubleshoot.

# ----------- Code with Classes (Recomended by ROS2) Template -----------
import rclpy 
from rclpy.node import Node               

class MyNode(Node):                                  # MyNode is the node name
    def __init__(self):
        super().__init__("py_test")                  # "py_test" is the node-name in the ros2 graph
        
        self.counter = 0                             # State variable to track execution ticks
        self.get_logger().info("Hello ROS2")         # Log node initialization status to stdout/rosout
        self.create_timer(1.0, self.timer_callback)  # Register a periodic timer behavior, Adds the timer to the global executor loop
        
    def timer_callback(self):                        # Timer allows to execute a command every n amount of time
        self.get_logger().info("Hello")
        self.counter += 1

def main(args=None):
    rclpy.init(args=args)                            # Initialize the ROS2 communications middleware layer (RMW)         
    node = MyNode()                                  # Instantiate the node executor target

    # It pauses the main script: It stops the code from reaching the end of the file so your program doesn't close.
    # It enters a "listening" loop: It constantly checks, "Is it time to run the timer yet?" or "Did we get a new message?"
    # It triggers your actions: The moment 1.0 second passes, spin() spots it and instantly runs your timer_callback function.
    rclpy.spin(node)                                 


    rclpy.shutdown()                                 # Context destruction: release ROS2 resources, destroy nodes, shutdown middleware

if __name__ == "__main__":                
    main()  

# Only prints text to its own terminal window. It does not broadcast any data across the ROS2 network. Other nodes cannot see or hear it.
# Therefore this next example includes the core ros2 functionality of publish/subscribe

import rclpy 
from rclpy.node import Node          
from std_msgs.msg import String              # Import standard ROS2 primitive message types

class MyNode(Node):                                 
    def __init__(self):
        super().__init__("py_test")                  
        
        self.counter = 0                             
        self.get_logger().info("Hello ROS2")         

        # ---------- Publisher ----------

        # Publisher: Creates a pipe named "/example_topic" using String messages.
        # Queue size (n) buffers outgoing messages if network latency peaks.
        self.publisher_ = self.create_publisher(String, "example_topic", 10)

        # ---------- Subscriber ----------

        # Subscriber: Listens to "/example_topic". 
        # Whenever a message arrives, the executor interrupts to run 'listener_callback'.
        self.subscription = self.create_subscription(
            String, 
            "example_topic", 
            self.listener_callback, 
            10
        )

        # Note; A package can have its node subscribe to n amount of topics, and pubish to m amount of topics

        self.create_timer(1.0, self.timer_callback)  
        
    def timer_callback(self):                     

        # Construct the standard ROS2 message object
        msg = String()
        msg.data = f"Hello ROS2 Network: {self.counter}"

        # Broadcast the data across the active DDS network
        self.publisher_.publish(msg)

        self.get_logger().info("Hello")
        self.counter += 1

    def listener_callback(self, msg):
        # Asynchronous event handler triggered when data arrives on monitored topic
        self.get_logger().info(f"Subscribed Listener Heard: '{msg.data}'")

def main(args=None):
    rclpy.init(args=args)                                     
    node = MyNode()                                  
    rclpy.spin(node)                                 
    rclpy.shutdown()                               
if __name__ == "__main__":                
    main()  